Implement async rest api routes and centralized error handling

Nana-EC opened the pull request:
**Detailed description**:
Issue #631 noted a few short falls of The REST API code. Including 
- unused Cacher.js
- partial asynchronous flows in requests
- Tedious Error object sets in response with lots of distributed similar code
- Synchronous and asynchronous error catching mixed together 

This change addresses the asks by
- Adding @awaitjs/express to allow for a middleware error handling using express logic
- Making all route calls asynchronous 
- Moving error logic to centralized error.js
- Bubbling errors up to middleware
- Removing unused methods from utils.js
- Updating validation methods to throw errors instead of setting response
- Updating specs and tests appropriately, also utilizing jest snapshots

**Which issue(s) this PR fixes**:
Fixes #631 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated

apeksharma commented on hedera-mirror-rest/server.js:
```diff
@@ -40,7 +41,7 @@ const accounts = require('./accounts.js');
 const topicmessage = require('./topicmessage.js');
 const eventAnalytics = require('./eventAnalytics.js');
 const utils = require('./utils.js');
-const Cacher = require('./cacher.js');
```
events and eventAnalytics are also unused. We can remove them.

steven-sheehy replied:
Those should be handled in #572

apeksharma commented on hedera-mirror-rest/helpers/error.js:
```diff
@@ -0,0 +1,88 @@
+    return;
+  }
+
+  if (statusCode === httpStatusCodes.INTERNAL_ERROR) {
```
Where is this thrown? Can't seem to find the code which throws this.
Either ways, let's add the comment explaining the need to override the message..."do not expose detailed internal errors externally...etc etc"

Nana-EC replied:
By default when you throw any error Express sets it to to a 500
See https://expressjs.com/en/guide/error-handling.html "The default error handler" section
But yeah good suggestion on the comment

Nana-EC replied:
I Amy be wrong or something else is happening. Thought according to that documentation that 500 would be the default, turns out in some cases its undefined, so setting that also

Nana-EC replied:
Updated to handle undefined and default cases where express may set the default http status code on an error

apeksharma commented on hedera-mirror-rest/topicmessage.js:
```diff
@@ -60,37 +63,33 @@ const validateGetSequenceMessageParams = function (topicId, seqNum) {
-  let badParams = [];
   if (!utils.isValidEntityNum(topicId)) {
-    badParams.push(utils.getInvalidParameterMessageObject(topicMessageColumns.TOPIC_NUM));
+    throw new ErrorHandler(httpStatusCodes.BAD_REQUEST, [
```
This seems to be popular error. Would be good to have a function in `error.js` like
```
const ErrorForInvalidParams = function (params) => {
  messages = params.map(param => {
    return {message: `Invalid parameter: ${parama}`}
  })
  return new ErrorHandler(BAD_REQUEST, messages);
}
```
Then `badParams` will actually be just params, rather than messages.
Will also remove many calls to `getInvalidParameterMessageObject` and utils will have 2 less functions.

steven-sheehy replied:
Agree, except would recommend a concrete `InvalidArgumentError extends Error` that does this.

Nana-EC replied:
I'm creating an InvalidArgumentError class that encapsulates this logic

apeksharma commented on hedera-mirror-rest/topicmessage.js:
```diff
@@ -60,37 +63,33 @@ const validateGetSequenceMessageParams = function (topicId, seqNum) {
   }
 
-  return utils.makeValidationResponse(badParams);
+  return true;
```
is return value still needed?

Nana-EC replied:
Just for testing purposes. No product code consumes it. It's true or throw error.

Nana-EC replied:
Will look into jest .not.throwError() and then remove the returns

apeksharma commented on hedera-mirror-rest/topicmessage.js:
```diff
@@ -271,31 +253,33 @@ const getMessages = async (pgSqlQuery, pgSqlParams) => {
-const getTopicMessages = (req, res) => {
+/**
+ * Handler function for /topics/:id API.
+ * @param req
```
nit: empty tags can be removed

Nana-EC replied:
Updated comment

apeksharma commented on hedera-mirror-rest/transactions.js:
```diff
@@ -304,30 +297,26 @@ const getOneTransaction = function (req, res) {
+      throw new ErrorHandler(httpStatusCodes.NOT_FOUND, 'Not found');
+    }
+
+    if (process.env.NODE_ENV === 'test') {
```
we can make this just logger.debug() too. If okay to have debug as default log level in tests, but tests hooks in prod code are code smell.
i know this is not your code, but I blame you for making all the past code here awesome.

steven-sheehy replied:
Agree, don't like `if (process.env.NODE_ENV === 'test')` anywhere. Think this is not used for logging though but for unit tests. But code should probably be refactored later to move sql construction to separate method that can be unit tested. Think that's a bigger change though.

Nana-EC replied:
Agreed. Nice thing is it no longer seems to be utilized, logic was probably in some tests I cleaned up in the past.
Removing this now.

Nana-EC replied:
There's another instance of this that does break test code when removed. Will address this in another PR like Steven noted. Topicmessage.js is more broken up where you can test for the sql query easier. When that similar refactor happens this won't be needed

steven-sheehy commented on hedera-mirror-rest/helpers/error.js:
```diff
@@ -0,0 +1,88 @@
+
+  const {statusCode, message} = err;
+
+  if (/ECONNREFUSED/.test(message)) {
```
It's not possible to catch as it appears as a generic `Error` object, so we have to interrogate the message

steven-sheehy commented on hedera-mirror-rest/server.js:
```diff
@@ -91,35 +92,25 @@ app.use(bodyParser.json());
 
-app.get(apiPrefix + '/topics/:id', topicmessage.getTopicMessages);
-app.get(apiPrefix + '/topic/:id', topicmessage.getTopicMessages);
+// app.use(utils.errorHandler);
```
Can remove

Nana-EC commented on hedera-mirror-rest/helpers/error.js:
```diff
@@ -0,0 +1,88 @@
+ */
+'use strict';
+
+class ErrorHandler extends Error {
```
Good suggestion. I can go with HttpError since it's just a plain Error object like you pointed out. The handleError handles the handling 😀

Nana-EC commented on hedera-mirror-rest/server.js:
```diff
@@ -91,35 +92,25 @@ app.use(bodyParser.json());
 app.use(compression());
 app.use(cors());
 
-let caches = {};
```
Yeah. The issue requests the Cacher.js to be removed as the internal logic is actually not fully implemented. I opened #668 to address whether there's still an option for saving it or if it should be removed and or replaced with something else

steven-sheehy commented on hedera-mirror-rest/helpers/error.js:
```diff
@@ -0,0 +1,88 @@
+  }
+}
+
+const httpStatusCodes = {
```
I don't like the intermixing of three separate concerns:

1. Error itself
2. Error handling
3. HTTP

There should not be a single error type but a hierarchy of errors like InvalidArgumentError, NotFoundError, etc. These should not contain any HTTP specific knowledge like status codes. There should be a HttpErrorHandler that maps these application errors to HTTP responses.

Nana-EC replied:
Makes sense. 
Though I think there's an argument to create an httpError type that implements Error and the rest inherit, therefore allowing each Error to be aware of its own status code. In which case there's no need for the mapper, just a parser like the handleError() which picks out the status code and message when creating the response.

Nana-EC replied:
I went with
1. Customer error types that contain their message logic
2. Separate error handling method that manages http vs db vs other errors
3. HTTP codes and messages are function in an HTTPError file. Customer errors as well as handler can pull code and message from here

I think this is better but let me know if there's anything else you feel strongly about

steven-sheehy commented on hedera-mirror-rest/helpers/error.js:
```diff
@@ -0,0 +1,88 @@
+    return;
+  }
+
+  if (statusCode === httpStatusCodes.INTERNAL_ERROR) {
```
We shouldn't be throwing INTERNAL_ERROR. Instead, it should be the default response if the error is not one of the errors we explicitly throw.

Nana-EC replied:
We don't throw it, I thought express would default to this. This was meant to catch any accidental errors and makes sure the message sent out doesn't expose internal code through stack trace.

Upon further testing I realize the status code is undefined so will update that.

steven-sheehy commented on hedera-mirror-rest/cacher.js:
```diff
@@ -50,17 +50,13 @@ class Cacher {
-      .catch((err) => {
-        utils.errorHandler(err, req, res, null);
-      });
+    func(req, res).then((data) => {
```
Let's just remove the file. This logic can move to the error handler

Nana-EC replied:
Will do

steven-sheehy commented on hedera-mirror-rest/balances.js:
```diff
@@ -119,7 +113,7 @@ const getBalances = function (req, res) {
 
   // Execute query
-  return pool.query(pgSqlQuery, sqlParams).then((results) => {
+  return await pool.query(pgSqlQuery, sqlParams).then((results) => {
```
Why are we blocking here? Should just return the promise. Same comment everywhere else.

Nana-EC replied:
Removed awaits

steven-sheehy commented on hedera-mirror-rest/server.js:
```diff
@@ -91,35 +92,25 @@ app.use(bodyParser.json());
 
-let caches = {};
-for (const api of [
-  {name: 'transactions', ttl: config.api.ttl.transactions},
```
These options should also be removed from `config/application.yaml` and `docs/configuration.md`

Nana-EC replied:
Will do

steven-sheehy commented on hedera-mirror-rest/server.js:
```diff
@@ -40,7 +41,7 @@ const accounts = require('./accounts.js');
 const eventAnalytics = require('./eventAnalytics.js');
 const utils = require('./utils.js');
-const Cacher = require('./cacher.js');
+const {handleError} = require('./helpers/error');
```
Would prefer an `error` package and handler and concrete errors all existing in there. Helpers is generic and is not self descriptive

Nana-EC replied:
Makes sense. Will do

steven-sheehy commented on hedera-mirror-rest/topicmessage.js:
```diff
@@ -40,11 +41,13 @@ const columnMap = {
   }
-  return utils.makeValidationResponse(badParams);
+
+  return true;
```
All these validate methods should not need to return booleans anymore if they throw errors as Appy indicated below as well.

Nana-EC replied:
Removed

steven-sheehy commented on hedera-mirror-rest/transactions.js:
```diff
@@ -198,21 +199,16 @@ const reqToSql = function (req) {
 
   let query = reqToSql(req);
 
   logger.trace('getTransactions query: ' + query.query + JSON.stringify(query.params));
```
Don't have to do it in this PR, but we should probably wrap these heavyweight query logs in `if (logger.isTraceEnabled())` to avoid the performance penalty of constructing the strings.

Nana-EC replied:
Will do in future PR

steven-sheehy commented on hedera-mirror-rest/transactions.js:
```diff
@@ -240,10 +236,7 @@ const getTransactions = function (req) {
-      code: utils.httpStatusCodes.OK,
-      contents: ret,
-    };
+    res.json(ret);
```
Can we just return `ret` and higher level middleware convert to json? We may want to support other content types in the future, for example and would like to move service layer to not be http specific. Same comment applies to other areas.

Nana-EC replied:
Good idea. Create d a responseHandler to do this. Moved it and the errorHandler to a middleware folder.

steven-sheehy commented:
Did we test that this fixes the error in the original issue?

Nana-EC commented:
> Did we test that this fixes the error in the original issue?

Yep. It handles ECONNREFUSED related issues

steven-sheehy commented:
> > Did we test that this fixes the error in the original issue?
> 
> Yep. It handles ECONNREFUSED related issues

Eh, the original issue is not a ECONNREFUSED. I added the ECONREFUSED logic and opened the ticket.

apeksharma commented:
> Yeah. The issue requests the Cacher.js to be removed 

Ahh, i see. Seeing 'unused Cacher' in description, i misunderstood that it was about using the cacher.

On topic of error handling, i believe this PR is net +ve contribution in improving error handling. Am okay if further improvements to error handling are punted to later time.

I don't mind dead caching code and configs around if the plan is to do #668  soon. Otherwise we can remove them.

Nana-EC commented on hedera-mirror-rest/helpers/error.js:
```diff
@@ -0,0 +1,88 @@
+};
+
+const handleError = (err, req, res, next) => {
+  if (process.env.NODE_ENV !== 'test') {
```
Will do, this was just to reduce logs during local runs cause it's hard to decipher issues

apeksharma commented on hedera-mirror-rest/__tests__/utilsFilters.test.js:
```diff
@@ -211,16 +194,7 @@ describe('utils validateAndParseFilters tests', () => {
-      constants.filterKeys.ACCOUNT_PUBLICKEY,
-      constants.filterKeys.ACCOUNT_BALANCE,
-    ]);
+    verifyInvalidFilters(filters);
```
yeah, expected values broken into separate file decreases readability. We are going towards less cohesion and more coupling (between files, matching on test names) when the tests are simple enough to not need jest snapshots.

Also, this is an internal function (not api) with an internal object as expected value. Standard way of building expected value (using Error's constructor) and then matching the objects seems right.

Nana-EC replied:
Not sure I agree. 
The benefit of a snapshot is that you capture all the properties of the object and check it all at once. The tests don't have to have lines of expected object creation and you don't have to have lines of comparisons.

In the valid cases I do that comparison cause it's easier. In this case with the throwing of a detailed error, capturing its details isn't as easy as checking a single return type .

Nana-EC commented:
> > > Did we test that this fixes the error in the original issue?
> > 
> > 
> > Yep. It handles ECONNREFUSED related issues
> 
> Eh, the original issue is not a ECONNREFUSED. I added the ECONREFUSED logic and opened the ticket.

Gotcha. Just tested locally and found another potential db error when db is shutting down or starting up - 
I've moved regex matching for the 3 cases to a dbErorr class

Nana-EC commented on hedera-mirror-rest/__tests__/__snapshots__/topicmessage.test.js.snap:
```diff
@@ -0,0 +1,74 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`topicmessage validateConsensusTimestampParam tests Verify validateConsensusTimestampParam returns correct result for -1234567890.000000001 1`] = `
```
Maybe I'm missing something. Since the methods don't actually return a valid value it's not trivial to compare and ensure the error object has the desired format. If it was just a simple Error I'd agree, but with our format it's not.

When I tried the approach you noted for the first case this is what happens 
    Expected message: {"_status": {"messages": [{"message": "Invalid parameter: consensus_timestamp"}]}}
    Received value:   [Error: [object Object]]

I wasn't able how to force stringily of the received error by the system. Since it's not a return value but rather a throw I haven't been able to go deep enough to get it.

Additionally snapshots make situations like this super easy. You don't have to worry about getting things and parsing then and creating the exact matching object in the test. Rather you run it, confirm the return is as expected and lock that in. Any change even a space gets caught, where as coding out every property or line can result in you missing something.

let me know if you find a way to do what you suggested here, I'm probably just missing it - https://jestjs.io/docs/en/expect#tothrowerror

apeksharma replied:
You're right, it was throwing that error at first. Got it working with some tweaking.
https://github.com/hashgraph/hedera-mirror-node/commit/dbf7ece757e975fc854b796d8186a23a4585dbc6

The drawbacks of snapshot is:
- fragmentation of logic: Setup for actual and expected values in two places. Less cohesion. Also hurts readability.
- Duplicate declarations of param names since constants.filterKeys can't be used in snapshot file.

Nana-EC commented on hedera-mirror-rest/errors/formattedError.js:
```diff
@@ -0,0 +1,63 @@
+'use strict';
+
+class FormattedError extends Error {
+  constructor(message) {
```
Thanks, for the suggestion. Will incorporate some of it.

Nana-EC replied:
Took the message formatting and simplification. Updated most params to be something other than message.
Didn't want to change the main message property since by convention Error.message is a thing.

steven-sheehy commented:
>> Only HttpErrorHandler should know which custom error maps to which http error code, and how to format the error to return to user.

Hehe, that's what I said earlier

steven-sheehy commented on hedera-mirror-rest/topicmessage.js:
```diff
@@ -271,31 +246,33 @@ const getMessages = async (pgSqlQuery, pgSqlParams) => {
  * @return {Promise} Promise for PostgreSQL query
  */
-const getMessageByConsensusTimestamp = function (req, res) {
+const getMessageByConsensusTimestamp = async (req, res) => {
```
`res` can be removed from all the service layer methods.

Nana-EC replied:
IN the process of removing those left over

Nana-EC commented on hedera-mirror-rest/middleware/httpErrorHandler.js:
```diff
@@ -0,0 +1,55 @@
+
+  const {statusCode, message} = err;
+  // catch DB errors
+  if (DbError.isDbConnectionError(message)) {
```
Yeah. The hope/intention was that in future we'd be able to throw this error  like the others. In this case where we're not able to catch the case early on and throw it I'm making use of the regex matching.
Figured keeping it in the DbError class made sense instead of spreading DbError logic around the code base.

I agreed on the actual message used to determine if connection issue. Problem is I didn't want that class being used for non db errors. A DB error that isn't connection related should still probably return this issue.

I think I'll revert one of my earlier changes and just have a catch on the pg.query so we aren't guessing if it's a DB error.

Nana-EC commented on hedera-mirror-rest/middleware/httpErrorHandler.js:
```diff
@@ -0,0 +1,55 @@
+const {httpStatusCodes, HttpError} = require('../errors/httpError');
+
+const handleError = (err, req, res, next) => {
+  // only logs in non test environment
```
Updated

Nana-EC commented on hedera-mirror-rest/transactions.js:
```diff
@@ -265,8 +259,8 @@ const getOneTransaction = function (req, res) {
-    res.status(utils.httpStatusCodes.BAD_REQUEST).send(utils.createSingleErrorJsonResponse(message));
-    return;
+
+    throw new BadRequestError(message);
```
The real difference is InvalidArgument encapsulates the required formatting of 'Invalid parameter: '.
Not separating can make it confusing to distinguish between a parameter of say 'topic_num' which should be formatted and the above case which shouldn't be formatted. Even more so when a set of them are passed through.
Status code should have an impact here any longer since the errors are void of that logic.

apeksharma replied:
I agree that having two classes here for same logical error is not required. An earlier suggestion to avoid the same was making InvalidArgumentError class ctor general (and similar to other errors), and a helper method for common case. Copying it below. The code here will use the ctor. 

```
// Plain ctor like other errors. If any api wants to throw custom error messages on invalid argument, it can directly use ctor.

// factory method to help common case
static forParams(badParams) {
  if(!Array.isArray(badParams) {
    badParam = [badParams];
  }
  return new InvalidArgumentError(badParams.map(p => `${httpErrorMessages.INVALID_ARGUMENT}${message}`));
}

Nana-EC replied:
Will remove BadRequest and utilize Appy's suggestion

apeksharma commented on hedera-mirror-rest/errors/dbError.js:
```diff
@@ -0,0 +1,53 @@
+   * @returns {boolean}
+   */
+  isDbConnectionError(errorMessage) {
+    if (
```
can do `return (/ECONNREFUSED/.test(errorMessage) ||
      /Connection terminated unexpectedly/.test(errorMessage) ||
      /unable to read data from DB/.test(errorMessage))`

Nana-EC replied:
Thanks

Nana-EC commented on hedera-mirror-rest/middleware/httpErrorHandler.js:
```diff
@@ -0,0 +1,55 @@
+    logger.error(`Error processing ${req.originalUrl}: `, err);
+  }
+
+  const {statusCode, message} = err;
```
Fixed

Nana-EC commented on hedera-mirror-rest/config/application.yml:
```diff
@@ -1,21 +1,16 @@
-      name: mirror_node
-      port: 5432
-    shard: 0
+    mirror:
```
Fixed

Nana-EC replied:
Specified it in the package.json. We can tweak as need in future PR's

apeksharma commented on hedera-mirror-rest/errors/formattedError.js:
```diff
@@ -0,0 +1,36 @@
+ */
+'use strict';
+
+class FormattedError extends Error {
```
FormattedError is just converting single message to array. It is not a 'type of error'.
We should remove this and let `httpErrorHandler.errorMessageFormat` do the `if(!Array...)`

Nana-EC replied:
Good point. Missed this.

Nana-EC commented on hedera-mirror-rest/accounts.js:
```diff
@@ -173,11 +171,7 @@ const getAccounts = function (req) {
-      code: utils.httpStatusCodes.OK,
-      contents: ret,
-    };
+    req[constants.responseDataLabel] = ret;
```
I looked around online and saw the recommendation of using the req.
Res.locals does seem better. I thought touching the res would result in some of the errors I'd been seeing that only let you set the headers once.

I did test with the next() and it seems fine, I think it would be required if you had middleware that needed to do some more in depth logic following this.
I see what you're saying in the server.js.
I just Didn't want to make it bigger cause I also eventually want to clean up that file and move each of those routes out to their own files.

We'll need to do further refactoring to decouple the service layer from http, cause you won't want to pass in the req object either. Probably an object that contains the url. I'll leave that for later

Nana-EC replied:
I opened a ticket #677 to address the work I noted in the later part that we both want done

steven-sheehy commented on hedera-mirror-rest/errors/invalidArgumentError.js:
```diff
@@ -0,0 +1,42 @@
+
+module.exports = {
+  InvalidArgumentError,
+  InvalidArgumentErrorMessageFormat,
```
nit: Don't think we need to export InvalidArgumentErrorMessageFormat

Nana-EC replied:
Done

steven-sheehy commented on hedera-mirror-rest/errors/notFoundError.js:
```diff
@@ -0,0 +1,33 @@
+
+module.exports = {
+  NotFoundError,
+  NotFoundErrorMessage,
```
Same

Nana-EC replied:
Done

steven-sheehy commented on hedera-mirror-rest/middleware/httpErrorHandler.js:
```diff
@@ -0,0 +1,85 @@
+
+module.exports = {
+  handleError,
+  httpStatusCodes,
```
Same

Nana-EC replied:
Done

steven-sheehy commented on hedera-mirror-rest/topicmessage.js:
```diff
@@ -304,6 +292,7 @@ module.exports = {
   getMessageByConsensusTimestamp: getMessageByConsensusTimestamp,
   getMessageByTopicAndSequenceRequest: getMessageByTopicAndSequenceRequest,
   getTopicMessages: getTopicMessages,
+  topicMessageColumns: topicMessageColumns,
```
Same

Nana-EC replied:
Done

apeksharma commented:
There's one remaining comment. I spent some time figuring it out. Would be good to have it addressed too.
